test/integration-tests/Seaport.test.ts

```diff
@@ -141,4 +157,187 @@ describe('Seaport', () => {
+        params: [ALICE_ADDRESS],
+      })
+      alice = await ethers.getSigner(ALICE_ADDRESS)
+      permit2 = (await deployPermit2()).connect(alice) as Permit2
+      router = (await deployUniversalRouter(permit2)).connect(alice) as UniversalRouter
+      planner = new RoutePlanner()
+      tubbyCats = new ethers.Contract(TUBBY_ADDRESS, ERC721_ABI) as ERC721
+      weth = new ethers.Contract(WETH.address, ERC20_ABI) as ERC20
+
+      // send 1 eth from alice to router.address for weth approval
+      await alice.sendTransaction({ to: router.address, value: ethers.utils.parseEther('1.0') })
+      expect(await ethers.provider.getBalance(router.address)).to.eq(ethers.utils.parseEther('1.0'))
+      // max approve conduit for weth
+      const routerSigner = await ethers.getImpersonatedSigner(router.address)
+      await weth.connect(routerSigner).approve(OPENSEA_CONDUIT, ethers.constants.MaxUint256)
+      await weth.connect(routerSigner).allowance(router.address, OPENSEA_CONDUIT)
+      expect(await weth.connect(routerSigner).balanceOf(router.address)).to.eq(0)
+    })
+
+    it('accepts an advanced order with criteria offering WETH', async () => {
+      // https://etherscan.io/tx/0x74551f604adea1c456395a8e801bb063bbec385bdebbc025a75e0605910f493c
+      let { advancedOrder, criteriaResolvers } = getAdvancedOrderParams(seaportOrders[2])
+      const value = calculateValue(advancedOrder.parameters.consideration, [ItemType.ERC20])
+      const params = advancedOrder.parameters
+      // Can add logic to select approval target (conduit or consideration) depending on conduitHash
+      // TODO: add helper function to get identifier from criteriaResolvers if exists
+      const id = criteriaResolvers[0].identifier
+      // transfer nft to alice as tubbyCatOwner
+      const prevTubbyCatOwner = await tubbyCats.connect(alice).ownerOf(id)
+      await tubbyCats
+        .connect(await ethers.getImpersonatedSigner(prevTubbyCatOwner))
+        .transferFrom(prevTubbyCatOwner, alice.address, id)
+
+      const wethReceived = BigNumber.from(advancedOrder.parameters.offer[0].startAmount).sub(value)
+
+      tubbyCats = tubbyCats.connect(alice)
+      const ownerBefore = await tubbyCats.ownerOf(id)
+      expect(ownerBefore).to.eq(alice.address)
+
+      const calldata = seaportInterface.encodeFunctionData('fulfillAdvancedOrder', [
+        advancedOrder,
+        criteriaResolvers,
+        OPENSEA_CONDUIT_KEY,
+        ADDRESS_ZERO, // 0 addr so router custody
+      ])
+
+      // TODO: need to add arg for msg.value?
+      planner.addCommand(CommandType.SEAPORT_SELL_721, [
+        calldata,
+        tubbyCats.address,
+        OPENSEA_CONDUIT,
+        id,
+        alice.address,
+      ])
+      planner.addCommand(CommandType.SWEEP, [WETH.address, MSG_SENDER, 0])
+      const { commands, inputs } = planner
+
+      const wethBefore = await weth.connect(alice).balanceOf(alice.address)
+      expect(wethBefore).to.eq(0)
```

Comment:
I'm personally not a fan of having expect statements scattered everywhere (though I see it's happened in some files that I haven't worked on :P)

If they must be scattered, I'd rather them be before and after the call we're testing. Ideally EVERY test is testing ONE CALL ONLY. Maybe it's fine to have a list of expects before the call, and a list after. I usually put all my expectations at the bottom of the test so the SPEC IS EXTRMELY CLEAR TO THE READER and they don't have to look all over the file/test. This specific expect doesn't look necessary as later on we're doing `expect(wethAfter.sub(wethBefore)).to.eq(wethReceived)` so why does it need to be 0 to begin with?

Reply:
Removed the 0 check, and yeah I think I'm going to refactor to use covens and only use a specific order / txn so I can move all of the nft approval stuff into the beforeEach, and only have the seaport call in the main test body

Reply:
I agree with emily on the point that I would like all the expects gathered into 2 blocks
- expects immediately before the router call
- expects immediately after the router call

In general I prefer to err on the side of more instead of less (differently to emily). If adding an expect means you are more certain that the expected behaviour is occurring then I'd say add it - it increases the reliability and security of your tests.

However I do agree that we dont actually care _what_ alice's balance is before the test - we just care that it goes up by the expected amount.

Reply:
Sounds good, changed in latest commit to have all of the expects at the very end of the test case, and `before` vars strictly before the call to the router, and all `after` vars strictly after.

Reply:
I don't think I was saying that less expects are better in general. Just that there shouldn't be expects testing things other than the call we're testing.